mdformat/_util.py

```diff
@@ -1,21 +1,81 @@
-import re
-from typing import Iterable
+from html.parser import HTMLParser
+from typing import Iterable, List, Optional, Set
 
 from markdown_it import MarkdownIt
 
+import mdformat.plugins
 
-def is_md_equal(md1: str, md2: str, *, ignore_codeclasses: Iterable[str] = ()) -> bool:
+
+def is_md_equal(
+    md1: str,
+    md2: str,
+    *,
+    enabled_extensions: Iterable[str] = (),
+    enabled_codeformatters: Iterable[str] = (),
+) -> bool:
     """Check if two Markdown produce the same HTML.
 
-    Renders HTML from both Markdown strings, strips whitespace and
-    checks equality. Note that this is not a perfect solution, as there
-    can be meaningful whitespace in HTML, e.g. in a <code> block.
+    Renders HTML from both Markdown strings, strip content of tags with
+    specified classes, and checks equality.
     """
-    html1 = MarkdownIt().render(md1)
-    html2 = MarkdownIt().render(md2)
-    html1 = re.sub(r"\s+", "", html1)
-    html2 = re.sub(r"\s+", "", html2)
-    for codeclass in ignore_codeclasses:
-        html1 = re.sub(rf'<codeclass="language-{codeclass}">.*</pre>', "", html1)
-        html2 = re.sub(rf'<codeclass="language-{codeclass}">.*</pre>', "", html2)
+    ignore_classes = [f"language-{lang}" for lang in enabled_codeformatters]
+    mdit = MarkdownIt()
+    for name in enabled_extensions:
+        plugin = mdformat.plugins.PARSER_EXTENSIONS[name]
+        plugin.update_mdit(mdit)
+        ignore_classes.extend(getattr(plugin, "ignore_classes", []))
+    html1 = HTML2JSON().parse(mdit.render(md1), ignore_classes)
+    html2 = HTML2JSON().parse(mdit.render(md2), ignore_classes)
+
     return html1 == html2
+
+
+class HTML2JSON(HTMLParser):
```

Comment:
This seems to output a syntax tree, not JSON. Maybe rename to `HTML2AST` or similar?

Reply:
Well yes it does produce valid JSON, which also happens to be an AST of sorts. So really either is apt.
Note that the other approach here is just to use https://www.crummy.com/software/BeautifulSoup/bs4/doc/ (that's the code I originally wrote), i.e. having an extra (albeit very well known) dependency versus using a bespoke parser that requires a little more validation testing. I'll leave you to decide which?

Reply:
I believe JSON is a text format, so we'd need to do `json.dumps()` to the output before we can call it JSON.

I think this custom parser should be just fine if we can keep it reasonably compact.

Reply:
yep fair